gui/main.cpp

```diff
@@ -24,20 +31,61 @@ int main(int argc, char *argv[]) {
   parser.addOption(startupOnlyOption);
   parser.process(app);
 
-  auto generate_data = []() -> std::vector<std::tuple<float, float>> {
+  GuiStateContainer state_container(
+      /*history_window=*/std::chrono::milliseconds(30000));
+  auto startup_time = state_container.GetStartupTime();
+
+  std::unique_ptr<ConnectedDevice> device =
+      std::make_unique<FakeConnectedDevice>(
+          [&](const GuiStatus &gui_status,
+              ControllerStatus *controller_status) {
+            // For now, completely fake the data and ignore gui_status.
+            (void)gui_status;
+
+            controller_status->uptime_ms =
+                std::chrono::duration_cast<std::chrono::milliseconds>(
+                    std::chrono::steady_clock::now() - startup_time)
+                    .count();
+            auto *sensors = &controller_status->sensor_readings;
+            sensors->pressure_cm_h2o =
+                sin(controller_status->uptime_ms * 0.001);
+            sensors->volume_ml = 0;
+            sensors->flow_ml_per_min = 0;
+          });
+
+  // TODO: Bind the readable aspects of state_container to UI elements
+  // TODO: Bind the writable aspects of state_container to parameters set in the
+  // UI
+  PeriodicClosure maintain_history(std::chrono::milliseconds(10), [&]() {
+    ControllerStatus controller_status;
+    device->ExchangeState(state_container.GetGuiStatus(), &controller_status);
+    state_container.AppendHistory(controller_status);
+  });
+  maintain_history.Start();
+
+  QObject::connect(&app, &QApplication::aboutToQuit,
+                   [&]() { maintain_history.Stop(); });
+
+  auto generate_pressure = [&]() -> std::vector<std::tuple<float, float>> {
     std::vector<std::tuple<float, float>> res;
-    for (int i = 0; i < 1000; ++i) {
-      res.push_back({i, sin(0.01 * i)});
+    for (const auto &[time, controller_status] :
+         state_container.GetControllerStatusHistory()) {
```

Comment:
The main principle here is that I want zero risk of blocking the thread that talks over the protocol, so running a user-defined callback under lock is out of the question. The code definitely does too many copies now, I'm going to add a TODO to get rid of them in case it's actually too slow. And I think it will change when I (or somebody) address the other comment about filling in the state of all series at the same time.

Reply:
How about we just move this loop into the state_controller class then?

I'm not just worried about the speed of the copies, but also the large amount of malloc traffic.  I am not really sure how this is going to work on the rpi -- how do we test that the heap does not sloooowly become fragmented over the course of weeks?  That's a bigger question than just this PR, for sure.

Reply:
Sorry I'm not sure how to move this into state_container. state_container stores this stuff as (SteadyInstant, ControllerStatus) tuples, which I think is the right format to store it in. Then generate_data converts it to renderable form (eventually to several renderable forms) - (millis ago, float); doing this conversion in state_container would be expensive as millis_ago constantly changes. Then DataSource converts it to QT types for rendering, which is also hard to avoid. We could fuse the first and the second conversion together, but I'd rather do that in a separate PR and together with the refactoring to fill in all three graphs at the same time.